### PR TITLE
[ci] Add a CI-specific bazelisk wrapper

### PR DESCRIPTION
**Pull request by mcy**
This wrapper applies options that are broadly helpful for CI jobs, so
that CI jobs can just write ci/bazelisk.sh to have them all applied
automatically.

As a side-effect, the quick-bazel.sh script is no more. As a side-effect, this PR also fixes some serious latency issues around accidentally building Verilator, although the fix is a hack being tracked here: lowRISC/opentitan/issues/12259.

### CODE REVIEW

**Review by milesdai — APPROVED** [review]

File: ci/.bazelrc [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# If something goes wrong, show the whole command line for it.
+build --verbose_failures
+
+# If one action fails, keep going, so that CI runs show as much actionable information
+# as possible.
+build --keep_going
+
+# Use the remote Bazel cache for CI.
+build --remote_cache=https://storage.googleapis.com/opentitan-bazel-cache
+build --remote_timeout=600
```
Comment (rswarbrick) [inline ci/.bazelrc:23]: Missing newline

File: ci/bazelisk.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# This is the CI version of `bazelisk.sh`, which calls into the "usual" wrapper,
+# but adds various flags to produce CI-friendly output. It does so by prociding a
+# command-line specified .bazelrc (that is applied alongside //.bazelrc).
+
+# An additional bazelrc must be synthesized to specify precisely how to use the
+# GCP bazel cache.
+GCP_BAZELRC="$(mktemp /tmp/XXXXXX.bazelrc)"
```
Comment (rswarbrick) [inline ci/bazelisk.sh:12]: Probably better to use a trap to make sure this gets cleaned up properly.

File: ci/bazelisk.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    echo "No key/invalid path to key. Download from cache only."
+    echo "build --remote_upload_local_results=false" > "$GCP_BAZELRC"
+fi
+
+"$(dirname $0)"/../bazelisk.sh \
+  --bazelrc="$GCP_BAZELRC" \
+  --bazelrc="$(dirname $0)"/.bazelrc \
+  $@
+
+rm $GCP_BAZELRC
```
Comment (rswarbrick) [inline ci/bazelisk.sh:26]: Missing newline. (The second one: editor mis-configuration?)

File: hw/BUILD [hunk 4]
```diff
@@ -17,6 +17,37 @@ fusesoc_build(
...
+)
+
+# This is used in CI steps that do not want to run Verilator tests, and thus
+# do not want to accidentally build Verilator. This causes the Verilator
+# target to not emit any files, which will break any tests that actually rely
+# on this.
+config_setting(
+    name = "fake_fusesoc",
+    values = {"define": "DISABLE_VERILATOR_BUILD=true"},
+)
```
Comment (rswarbrick) [inline hw/BUILD:30]: I'm not really sure I understand the motivation here: if you don't want to run stuff that needs Verilator, can't you just not depend on it? This looks like the sort of cleverness that just confuses people down the line.

File: hw/BUILD [hunk 5]
```diff
@@ -17,6 +17,37 @@ fusesoc_build(
     data = ["//hw/ip/otbn:all_files"],
     systems = ["lowrisc:dv:chip_verilator_sim"],
     target = "sim",
+    visibility = ["//visibility:private"],
+)
+
+# This is used in CI steps that do not want to run Verilator tests, and thus
+# do not want to accidentally build Verilator. This causes the Verilator
```
Comment (rswarbrick) [inline hw/BUILD:24]: Probably worth getting the terminology right here: Verilator is a program that we depend on. We use it to build a verilated model of the chip.

**Comment by rswarbrick** [comment]
Also, this single commit does several different things. I'm clearly not in charge of reviewing the Bazelification work, so I guess maybe that's fine? But git histories are much nicer with smaller commits.

File: hw/BUILD [hunk 6]
```diff
@@ -17,6 +17,37 @@ fusesoc_build(
...
+# target to not emit any files, which will break any tests that actually rely
+# on this.
+config_setting(
+    name = "fake_fusesoc",
+    values = {"define": "DISABLE_VERILATOR_BUILD=true"},
+)
+
+genrule(
```
Comment (drewmacrae) [inline hw/BUILD:32]: This should be relocated.

File: azure-pipelines.yml [hunk 7]
```diff
@@ -597,6 +597,14 @@ jobs:
     - bash: echo "##vso[task.setvariable variable=bazelCacheGcpKeyPath]$(bazelCacheGcpKey.secureFilePath)"
       condition: eq(variables['Build.SourceBranchName'], 'master')
       displayName: Set the remote cache GCP key path
-    - bash: ci/scripts/quick-bazel.sh $(bazelCacheGcpKeyPath)
-      # The default timeout is too short to build verilator so let's skip it for the first pass
-      displayName: "Build and Test Software with Bazel (except //hw:verilator)"
+    - bash: |
+        # This command builds all software and runs all unit tests that run on
+        # the host. Additionally, it explicitly does not build Verilator,
```
Comment (drewmacrae) [inline azure-pipelines.yml:602]: we do have a target //hw:verilator which is a verilated top_earlgrey, it's not really Verilator, which is confusing...

File: ci/.bazelrc [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+# as possible, but only updating once every second.
+build --show_timestamps 
+build --show_progress_rate_limit=1
+
+# If something goes wrong, show the whole command line for it.
+build --verbose_failures
+
+# If one action fails, keep going, so that CI runs show as much actionable information
+# as possible.
+build --keep_going
```
Comment (drewmacrae) [inline ci/.bazelrc:19]: While we're frustrated with the load and latency on CI, and now that we have Bazel tests running clean and enforced by CI, --keep_going is going to mean spending more time waiting for CI to build PRs that need to be fixed.

**Review by drewmacrae — COMMENTED** [review]
I see a few topics here, the ci/bazelisk.sh work LGTM,
I'd like to see a precedent for the fake swapping instead of querying and excluding, If the querying is getting dependencies wrong, we should try to figure out why, and I'd like to not worry about what the experience of trying to test with a faked //hw:verilator target, especially if it's only likely to happen in CI where it's harder to reproduce and observe.

File: hw/BUILD (hunk 4 above)
Comment (cfrantz) [reply hw/BUILD:30]: Please create an issue to understand what is happening with the dependency graph with the goal of eliminating this `config_setting`.  Leave a TODO comment here referencing the issue.

**Review by cfrantz — APPROVED** [review]

File: ci/bazelisk.sh (hunk 3 above)
Comment (mcy) [reply ci/bazelisk.sh:26]: https://github.com/lowRISC/opentitan/pull/12257

**Review by mcy — COMMENTED** [review]
Apparently Github decided my comments would count as a code review, oops.

**Pull request merged**